### ISSUE TITLE
UMAP crashes when torch also imported before first run

### ISSUE THREAD

**Issue by willeppy** [lmcinnes/umap#1094]
Hi,

I am trying to use UMAP to reduce the dimension of torch embeddings, but I am experiencing weird behavior with crashes when torch and umap are both imported.

## Versions
Using a Mac with Intel processor
python 3.10.13
torch version 2.2.0
umap-learn version 0.5.5

## Example fail
The below code fails with unspecified seg fault:
```python
import umap
import torch


def get_projection(vector):
    return umap.UMAP(metric="euclidean", n_components=2, verbose=True).fit_transform(
        vector
    )

def run():
    embeddings = torch.load("./embeddings.pt").numpy()
    get_projection(embeddings)

run()
```
Gives output before failing
```
UMAP( verbose=True)
Tue Feb 20 15:17:15 2024 Construct fuzzy simplicial set
OMP: Info #273: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
Tue Feb 20 15:17:22 2024 Finding Nearest Neighbors
```

## Example succeed
The weird thing is if I run UMAP once before importing torch, it runs successfully so I think is something weird about importing both packages?
The below code works
```python
import umap
import numpy as np


def get_projection(vector):
    return umap.UMAP(metric="euclidean", n_components=2, verbose=True).fit_transform(
        vector
    )


def run():
    # run one before import
    np_easy = np.random.rand(10, 10)
    get_projection(np_easy)
    print("run 1 successful")

    import torch
    # run two after import now works??
    embeddings = torch.load("./embeddings.pt").numpy()
    get_projection(embeddings)
    print("run 2 successful")

run()
```

I suspect this might be some issue with conflicts between torch and umap but have been struggling to debug so would appreciate any help!

**Comment by lmcinnes** [comment]
That is a little baffling to me to be honest. To just crash like that I think it needs to be related to something inside numba compiled functions interacting strangely. It may be related to threading controls?

Some things to try:

* See if you can import and run pynndescent on your dataset (``pynndescent.NNDescent(vector)``) or whether that crashes in the same way depending on torch imports.
* If that crashes try it with n_jobs=1 so that it runs single threaded (```pynndescent.NNDescent(vector, n_jobs=1)``) and see if it still crashes.
* If that crashes try setting some threading control options for numba (see https://numba.pydata.org/numba-doc/dev/reference/envvars.html#threading-control and see if any of those make a difference)
* If that crashes try disabling numba JIT altogether (set the environment variable ``NUMA_DISABLE_JIT`` to 1).
* If that crashes ... I'm at a loss.

**Comment by willeppy** [comment]
Sorry for my slow response and thanks for your help on this.

Running `pynndescent.NNDescent(vector)` has same behavior depending on import order with torch.

BUT it works for me if I set `n_jobs=1`! Still not sure what weirdness is going on but luckily this seems to fix it so will close this issue
